% Treebase: An R package for discovery, access and manipulation of online phylogenies 

1.  The TreeBASE portal is an important and rapidly growing repository of
    phylogenetic data. The R statistical environment has also become a
    primary tool for applied phylogenetic analyses that use this kind
    of data across a range of questions, from comparative evolution
    to community ecology to conservation planning.

2.  We have developed `treebase`, an open-source  (freely available from
    [http://cran.r-project.org/web/packages/treebase](http://cran.r-project.org/web/packages/treebase))
    for the R programming environment, providing simplified, programmatic and
    interactive access to phylogenetic data in the TreeBASE repository.

3.  We illustrate how this package creates a bridge between the
    repository and the rapidly growing ecosystem of R packages for
    phylogenetics that can reduce barriers to discovery and integration
    across phylogenetic research.

4.  We show how the `treebase` package can be used to facilitate replication
    of previous studies and testing of methods and hypotheses across
    a large sample of phylogenies, which may help make such 
    important practices more common.


#### Keywords 

R, software, API, TreeBASE, database, e-science

Introduction
============


Applications that use phylogenetic information as part of their analyses
are becoming increasingly central to both evolutionary and ecological
research. The exponential growth in genetic sequence data available
for all forms of life has driven rapid advances in the methods that
can infer the phylogenetic relationships and divergence times across
different taxa [@huelsenbeck2001b; @stamatakis2006; @drummond2007].
Once again the product of one field has become the raw data of the next.
Unfortunately, while the discipline of bioinformatics has emerged to
help harness and curate the wealth of genetic data with cutting edge
computer science, statistics, and internet technology, its counterpart
in evolutionary informatics remains “scattered, poorly documented,
and in formats that impede discovery and integration” [@parr2011a]. Our
goal when developing the `treebase` package is to provide steps to reduce
these challenges through programmatic and interactive access between the
repositories that store this data and the software tools commonly used
to analyze them.  The tools provided in the `treebase` package can make
discovery and analysis easier and more replicable, and facilitate scaling
analyses across an ever growing repository of potential phylogenetic data.


The R statistical environment [@rteam2012] has become a dominant
platform for researchers using phylogenetic data to address a
rapidly expanding set of questions in ecological and evolutionary
processes. These methods include, but are not limited to, ancestral
state reconstruction [@paradis2004; @butler2004], diversification
analysis [@paradis2004; @rabosky2006b; @harmon2008], identifying
trait dependent speciation and extinction rates, [@fitzjohn2010;
@goldberg2011; @stadler2011], quantifying the rate and tempo of
trait evolution [@butler2004; @harmon2008; @eastman2011], identifying
evolutionary influences and proxies for community ecology [@webb2008;
@kembel2010], connecting phylogeny data to climate patterns [@warren2008;
@evans2009b], and simulation of speciation and character evolution
[@harmon2008; @stadler2011a; @boettiger2012], as well as various
manipulations and visualizations of phylogenetic data [@paradis2004;
@schliep2010; @jombart2010; @revell2011]. A more comprehensive list
of R packages by analysis type is available on the phylogenetics taskview,
[http://cran.r-project.org/web/views/Phylogenetics.html](http://cran.r-project.org/web/views/Phylogenetics.html).
A few programs for applied phylogenetic methods are written for
environments outside the R language, incuding Java [@maddison2011],
MATLAB [@blomberg2003] and Python [@sukumaran2010] and online interfaces
[@martins2004].


TreeBASE ([http://treebase.org](http://treebase.org)) is an online
repository of phylogenetic data (e.g. trees of species, populations, or
genes) that have been published in a peer-reviewed academic journal,
book, thesis or conference proceedings
[@sanderson1994b; @morell1996]. The database can be searched through
an online interface which allows users to find a phylogenetic tree from
a particular publication, author or taxa of interest. TreeBASE provides
an application programming interface (API) that lets computer
applications make queries to the database. Our `treebase` package uses
this API to create a direct link between this data and the R language.
This has several immediate and important benefits:

1. *Data discovery.*  Users can leverage the rich higher-level programming
environment provided by the R language to better identify data sets
appropriate for their research by constructing queries for datasets that
match appropriate metadata requirements.

2. *Programmatic data access.* Many tasks that are theoretically made
possible by the creation of the TreeBASE repository are not pursued
because they would be too laborious for an exploratory analysis.
The ability to use programatic access across data sets to automatically
download and perform a systematic analysis using the rich set of tools
available in R opens up new avenues for research.

3. *Automatic updating*.  The TreeBASE repository is expanding rapidly.
The scriptable nature of analyses run with our `treebase` package
means that a study can be rerun on the latest version of the repository
without additional effort but with potential new information.



Programmatic Web Access
-----------------------

The Treebase repository makes data accessible by web queries through a 
RESTful (REpresentational State Transfer) interface, which supplies search
conditions aaddress URL.  The repository returns the requested data in 
XML (extensible markup language) format.  The `treebase` package uses the 
`RCurl` package to make queries over the web to the repository, and the 
`XML` package to parse the webpage returned by the repository into meaningful
R data objects [@rcurl, @xml].  While these querying and parsing functions
comprise most of the code provided in the `treebase` package, they are 
hidden from the end user who can interact with these rich data retrivial and 
manipulation tools to access data from these remote repositories in much 
the same way as data locally available on the users harddisk.  



Basic queries
--------------

``` {r libs, echo=FALSE}
    library(treebase)
    library(ggplot2)
````


The basic functions of the TreeBASE API allow search queries through two
separate interfaces. The `OAI-PMH` interface provides the metadata
associated with the publications from which the phylogenies have been
taken, while the `Phylo-WS` interface provides information and access to
the phylogenetic data itself. The `treebase` package allows these queries to be
made directly from R, just as a user would make them from the web browser.
This enables a user to construct more complicated filters than permitted by the web interface,
and allows the user to maintain a record of the queries they used to collect their data
as an R script. Scripting the data-gathering process helps reduce errors and assists in 
replicating the analysis later, either by the authors or other researchers [@peng2011a].


The `search_treebase` function is the heart of the `treebase` package.
Table 1 lists each of the types of queries available through the
`search_treebase` function.  This list can also be found in the function 
documentation through the R command `?search_treebase`.  
Any of the queries available on the web interface can now be made
directly from R, including downloading and importing the phylogeny into
the R interface. For instance, one can search for phylogenies containing
dolphin taxa, "Delphinus," or all phylogenies submitted by a given
author, "Huelsenbeck" using the R command

``` {r basicQueries, eval=FALSE }
    search_treebase("Delphinus", by="taxon")
    search_treebase("Huelsenbeck", by="author")
````

This function loads the matching phylogenies into R, ready for analysis.
The package documentation provides many examples of possible queries.



  search "by="     description
  ---------------  -----------------------------------------------------
  abstract         search terms in the publication abstract
  author           match authors in the publication
  subject          Matches in the subject terms
  doi              The unique object identifier for the publication 
  ncbi             NCBI identifier number for the taxon
  kind.tree        Kind of tree (Gene tree, species tree, barcode tree)  
  type.tree        Type of tree (Consensus or Single)
  ntax             Number of taxa in the matrix
  quality          A quality score for the tree, if it has been rated.  
  study            Match words in the title of the study or publication
  taxon            Taxon scientific name 
  id.study         TreeBASE study ID
  id.tree          TreeBASE's unique tree identifier (Tr.id)
  id.taxon         Taxon identifier number from TreeBase 
  tree             The title for the tree

  Table: Queries available in `search_treebase`.  The first argument is
  the keyword used in the query such as an author's name, while the second
  argument indcates the type of query (_i.e._ "author").



Data discovery in TreeBASE
==========================

Data discovery involves searching for existing data that meets
certain desired characteristics.  Such searches take advantage of
metadata, summary information describing the data entries provided in
the repository.  The web repository uses separate interfaces (APIs)
to access metadata describing the publications associated with the
data entered, such as the publisher, year of publication, etc., and
the a different interface to describe the metadata associated with an
individual phylogeny, such as the number of taxa or the kind of tree
(_e.g._ gene tree vs. species tree).  The `treebase` package can query
these individal sources of metadata seperately, but this information
is most powerful when used in concert -- allowing the construction of
complicated searches that cannot be automated through the web interface.
The `metadata` function updates a list of all available metadata from
both APIs and returns this information as an R `data.frame`.


```{r metadata_table}
meta <- metadata()
````

This facilitates identifying all potetnial data which meets certain
desired attributes.  For instance, if an analysis requires consensus
gene trees with 100 or more taxa appearing in the journals _Nature_
or _Science_, one can quickly identify candidate phylogenies using the
R's subsetting syntax

```{r metadata_subset}
meta[publisher %in% c("Nature", "Science") & ntaxa > 100 & kind == "Gene Tree",]
````


The fields provided by `metadata` are listed in Table II.  

  metadata field     description
  ---------------  -----------------------------------------------------
  Study.id         TreeBASE study ID
  Tree.id          TreeBASE's unique tree identifier 
  kind             Kind of tree (Gene tree, species tree, barcode tree)  
  type             Type of tree (Consensus or Single)
  quality          A quality score for the tree, if it has been rated.  
  ntaxa            Number of taxa in the matrix
  date             Year the study was published
  author           First author in the publication
  title            The title of the publication 

Table: Columns of metadata avialable from the `metadata` function 



From the length of this list we see that there are currently `r
prettyNum(length(unique(meta[["Study.id"]])))` published studies in
the database.  Metadata can also be used to reveal trends in the data
deposition which may be useful in identifying patterns or biases in
research or emerging potential types of data.  As a simple example, we
look at trends in which publishers are submitting the greatest number
of phylogenies to TreeBASE over time,


``` {r   journals }
    date <- meta[["date"]] 
    publisher <- meta[["publisher"]]
````

We plot the distribution of publication years for phylogenies deposited
in TreeBASE, color coding by publisher in Fig [fig:1].

``` {r dates, cache=FALSE, fig.cap="Histogram of publication dates by year, with the code required to generate the figure."}
  library(ggplot2) 
  ggplot(meta) + geom_bar(aes(date, fill = publisher)) + 
    opts(axis.text.x=theme_text(angle=90, hjust=1))
````

Typically we are interested in the metadata describing the phylogenies
themselves rather than just in the publications in which they appeared.
Phylogenetic metadata includes features such as the number of taxa
in the tree, a quality score (if available), kind of tree (gene tree,
species tree, or barcode tree) or whether the phylogeny represents a
consensus tree from a distribution or just a single estimate.


We can summarize how these `r dim(meta)[1]` trees break out by kind or
type  using the `phylo_metadata` function to extract the kind of tree
(gene/species/barcode) and type (single or consensus):


``` {r  kind, eval=FALSE }
table(meta[["kind"]], meta[["type"]])
````

``` {r  kindtable, results='asis', echo=FALSE }
output <-  table(meta[["kind"]], meta[["type"]])
xtable::xtable(output)
````


Accessing all phylogenies
-------------------------

For certain applications a user may wish to download all the available
phylogenies from TreeBASE. Using the `cache_treebase` function allows
a user to download a local copy of all trees. Because direct database
dumps are not available from treebase.org, this function has intentional
delays to avoid overtaxing the TreeBASE servers, and should be allowed
a full day to run.

``` {r buildcache, eval=FALSE }
treebase <- cache_treebase()
```

Once run, the cache is saved compactly in memory where it can be easily
and quickly restored. For convenience, the `treebase` package comes with
a copy already cached, which can be loaded into memory.


``` {r loadcache}
data(treebase)
````

Having access to both the metadata from the studies and from the
phylogenies in R lets us quickly combine these data sources in interesting
ways. For instance, with a few commands we can visualize how the number
of taxa on submitted phylogenies has increasing over time, Figure [fig:2].

``` {r   taxagrowth, fig.cap="Combining the metadata available from publications and from phylogenies themselves, we can visualize the growth in taxa on published phylogenies. Note that the maximum size tree deposited each year is growing far faster than the average number."}
ggplot(meta, aes(date, ntaxa)) + geom_point() + stat_smooth(aes(group = 1))
````



The promise of this exponential growth in the sizes of available
phylogenies, with some trees representing `r max(taxa$Ntaxa, na.rm=TRUE)` 
taxa motivates the more and more ambitious inference
methods being developed which require large trees to have adequate signal
[@boettiger2012; @fitzjohn2009; @beaulieu2012].


Reproducible computations
=========================


Reproducible research has become a topic of increasing concern in
recent years, and facilitating access to data and using scripts that
can replicate analyses can help lower barriers to the replication of
statistical and computational results [@schwab2000; @gentleman2004;
@peng2011b].  The `treebase` package facilitates this process, as we
illustrate in a simple example.

Consider the shifts in speciation rate identified by @derryberry2011
on a phylogeny of ovenbirds and treecreapers, which uses the methods
provided in the R package  `laser` [@rabosky2006b].  We will seek to
not only replicate the results, but also compare them against methods
presented in @stadler2011 and implemented in the package `TreePar`, which
permits speciation models that were not available to @derryberry2011 at
the time of their study.

Obtaining the tree
------------------

By drawing on the rich data manipulation tools available in R which may
be familiar to the large R phylogenetics community, the ` treebase `
package allows us to construct richer queries than are possible through
TreeBASE alone.

The most expedient way to identify the data uses the digital object identifer
(doi) at the top of most articles, which we use in a call to the 
`search_treebase` function, such as

``` {r doiquery} 
results <- search_treebase("10.1111/j.1558-5646.2011.01374.x", "doi") 
````

The search returns a list, since some publications can contain many trees.
In this case our phylogeny is in the first element of the list,


``` {r doiqueryresults} 
results[[1]] 
```` 

``` {r firstone, echo=FALSE, include=FALSE} 
derryberry <- results[[1]] 
````

Having successfully imported the phylogenetic tree corresponding to this
study, we can quickly replicate their analysis of which diversification
process best fits the data.  These steps can be easily implemented using
the phylogenetics packages we have just mentioned. As a comparison of
speciation models is not the focus of this paper, the complete code
and explanation for these steps is provided as an appendix.  Happily,
this analysis confirms the author's original conclusions, even when the
more general model of @stadler2011 is considered.




Analyses across many phylogenies
================================

Large scale comparative analyses that seek to characterize evolutionary
patterns across many phylogenies are increasingly common in phylogenetic
methods [_e.g._ @mcpeek2007; @phillimore2008; @mcpeek2008; @quental2010;
@davies2011a].  Sometimes referred to by their authors as meta-analyses,
these approaches have focused on re-analyzing phylogenetic trees collected
from many different earlier publications.  This is a more direct approach
than the traditional concept of meta-analysis where statistical results
from earlier studies are weighted by their sample size without being able to
access the raw data.  Because the identical
analysis can be repeated on the original data from each study, this
approach avoids some of the statistical challenges inherent in traditional
meta-analyses summarizing results across heterogeneous approaches.

To date, researchers have gone through heroic efforts simply to assemble
these data sets from the literature.  As described in @mcpeek2007; (emphasis added) 

> One data set was based on 163 published species-level molecular
phylogenies of arthropods, chordates, and mollusks.  A PDF format file
of each article was obtained, and a digital snapshot of the figure was
taken in Adobe Acrobat  7.0. This image was transferred to a PowerPoint
(Microsoft) file and printed on a laser printer. The phylogenies included
in this study are listed in the appendix. _All branch lengths were
measured by hand from these  printed sheets using dial calipers._

Despite the recent emergence of digital tools that could now facilitate
this analysis without mechanical calipers, [_e.g._ treesnatcher, @laubach2007],
it is  easier and less error-prone to pull properly
formatted phylogenies from the database for this purpose. Moreover,
as the available data increases with subsequent publications, updating
earlier meta-analyses can become increasingly tedious. Using `treebase`,
a user can apply any analysis they have written for a single phylogeny
across the entire collection of suitable phylogenies in TreeBASE,
which can help overcome such barriers to discovery and integration at
this large scale.  Using the functions we introduce aboved, we provide
in the Appendix a simple example that computes the gamma statistic of
@pybus2000, which provides an measure of when speciation patterns differ
from the popular birth-death model.  We see that many phylogenies fall
outside the distribution of the statistic expected under that model.



Conclusion
==========

While we have focused on examples that require no additional data
beyond the phylogeny, a wide array of methods combine this data with
information about the traits, geography, or ecological community of the
taxa represented. In such cases we would need programmatic access to
the trait data as well as the phylogeny. The Dryad digital repository
([http://datadryad.org](http://datadryad.org)) is an effort in this 
direction. While programmatic access to the repository is possible
through the `rdryad` package [@chamberlain2012], variation in data
formatting must first be overcome before similar direct access to 
the data is possible.  Dedicated databases such as FishBASE
([http://fishbase.org](http://fishbase.org)) may be another alternative,
where morphological data can be queried for a list of species using the
`rfishbase` package [@rfishbase]. The development of similar software
for programmatic data access will rapidly extend the space and scale of
possible analyses.

The recent advent of mandatory data archiving in many of the major
journals publishing phylognetics-based research [_e.g._ @fairbairn2010;
@piwowar2011; @whitlock2010], is a particularly promising development that
should continue to fuel the trend of submissions seen in Fig. 1. Accompanied
by faster and more inexpensive techniques of NextGen sequencing, and the
rapid expansion in phylogenetic applications, we anticipate this rapid
growth in available phylogenies will continue. Faced with this flood
of data, programmatic access becomes not only increasingly powerful but
an increasingly necessary way to ensure we can still see the forest for
all the trees.

Acknowledgements
================

CB wishes to thank S. Price for feedback on the manuscript, the
TreeBASE developer team for building and supporting the repository,
and all contributers to TreeBASE. CB is supported by a Computational
Sciences Graduate Fellowship from the Department of Energy under grant
number DE-FG02-97ER25308.

 
``` {r save, echo=FALSE }
    save(list=ls(), file="knit_treebase.rda")
````




Appendix
========

Reproducible computation: A diversification rate analysis
---------------------------------------------------------

Different diversification models make different assumptions 
about the rate of speciation, extinction, and how these rates may be changing
over time.  The authors consider eight different models, implemented in the 
laser package [@rabosky2006b]. This code fits each of the eight models to that
data:

``` {r   RR, tidy=FALSE  }
library(ape)
library(laser)
bt <- branching.times(derryberry)
models <- list(
  yule = pureBirth(bt),  
  birth_death = bd(bt),     
  yule.2.rate = yule2rate(bt),
  linear.diversity.dependent = DDL(bt),    
  exponential.diversity.dependent = DDX(bt),
  varying.speciation_rate = fitSPVAR(bt),  
  varying.extinction_rate = fitEXVAR(bt),  
  varying_both = fitBOTHVAR(bt))
````

Each of the model estimate includes an AIC score indicating the goodness of
fit, penalized by model complexity (lower scores indicate better fits)
We ask R to tell us which model has the lowest AIC score,

``` {r   extract_aic  }
aics <- sapply(models, function(model) model$aic)
best_fit <- names(models[which.min(aics)])
```` 

and confirm the result presented in @derryberry2011; 
that the `r best_fit` model is the best fit to the data.  


In this fast-moving field, new methods often become available between the 
time of submission and time of publication of a manuscript. For instance, 
the more sophisticated methods available in the more recent package, `TreePar`, 
introduced in @stadler2011 were not used in this study.

``` {r   TreeSimError, echo=FALSE  }
# Locale settings to be safe
Sys.setlocale(locale="C") -> locale
rm(list="locale") # clean up
````

We load the new method and format the data as its manual instructs us

``` {r   treepar  }
require(TreePar)
x <- sort(getx(derryberry), decreasing = TRUE)
````

The best-fit model in the laser analysis was a yule (net diversification
rate) models with two separate rates.  We can ask ` TreePar ` to see if
a model with more rate shifts is favored over this single shift,
a question that was not possible to address using the tools provided in
`laser`. The previous analysis also considers a birth-death model that 
allowed speciation and extinction rates to be estimated separately, but 
did not allow for a shift in the rate of such a model.  Here we consider
models that have up to 4 different rates in Yule models, (The syntax in
`TreeParr` is slightly cumbersome, the [[2]] indicates where this command
happens to store the output models.)

``` {r   treepar_yule, include=FALSE  }
yule_models <- bd.shifts.optim(x, sampling = c(1,1,1,1), 
  grid = 5, start = 0, end = 60, yule = TRUE)[[2]]
````


``` {r   treepar_yule_show, eval=FALSE  }
yule_models <- bd.shifts.optim(x, sampling = c(1,1,1,1), 
  grid = 5, start = 0, end = 60, yule = TRUE)[[2]]
````

  
We also want to compare the performance of models which 
allow up to four shifts and also 
estimate extinction and speciation separately:

``` {r   treepar_birthdeath, include=FALSE  }
birth_death_models <- bd.shifts.optim(x, sampling = c(1,1,1,1), 
  grid = 5, start = 0, end = 60, yule = FALSE)[[2]]
````

``` {r   treepar_birthdeath_show, eval=FALSE  }
birth_death_models <- bd.shifts.optim(x, sampling = c(1,1,1,1), 
  grid = 5, start = 0, end = 60, yule = FALSE)[[2]]
````

The models output by these functions are ordered by increasing number of shifts.  
We can select the best-fitting model by AIC score, which is slightly cumbersome 
in `TreePar` syntax.  First compute the AIC scores of both the `yule_models` and the 
`birth_death_models` we fitted above,

``` {r   aic_yule1  }
yule_aic <- sapply(yule_models, function(pars) 2 * (length(pars) - 1) + 2 * pars[1] )
birth_death_aic <- sapply(birth_death_models, function(pars) 2 * (length(pars) - 1) + 2 * pars[1] )
````

And ten generate a list identifying which model has the best (lowest) AIC score among the Yule models and 
which has the best AIC score among the birth-death models, 

```{r aic_yule2}
best_no_of_rates <- list(Yule = which.min(yule_aic), birth.death = which.min(birth_death_aic))
````

The best model is then whichever of these has the smaller AIC value.  

```{r aic_yule3}
best_model <- which.min(c(min(yule_aic), min(birth_death_aic)))
````


which confirms that the `r names(best_no_of_rates[best_model])` `r best_no_of_rates[best_model][[1]]`-rate  
model is still the best choice based on AIC score.  Of the eight models 
in this second analysis, only three were in the original set considered 
(Yule 1-rate and 2-rate, and birth-death without a shift), so we could by
no means have been sure ahead of time that a birth death with a shift, or
a Yule model with a greater number of shifts, would not have fitted better.  

Tests across many phylogenies
-----------------------------

A standard test of this is the gamma statistic of @pybus2000
which tests the null hypothesis that the rates of speciation and extinction
are constant. The gamma statistic is normally distributed about 0 for a pure
birth or birth-death process, values larger than 0 indicate that internal 
nodes are closer to the tip then expected, while values smaller than 0 indicate
nodes farther from the tip then expected.  In this section, we collect all 
phylogenetic trees from TreeBASE and select those with branch length data that 
we can time-calibrate using tools available in R.  We can then calculate the 
distribution of this statistic for all available trees, and compare these 
results with those from the analyses mentioned above.  

The `treebase` package provides a compressed cache of the phylogenies
available in treebase.  This cache can be automatically updated with the
`cache_treebase` function,

``` {r meta_harvest, eval=FALSE}
treebase <- cache_treebase()
````

which may require a day or so to complete, 
and will save a file in the working directory named with treebase and
the date obtained.  For convenience, we can load the cached copy 
distributed with the ` treebase ` package:

``` {r cachedcopy}
data(treebase)
````

We will only be able to use those phylogenies that include branch length data.
We drop those that do not from the data set, 

``` {r branchlengthfilter}
      have <- have_branchlength(treebase)
      branchlengths <- treebase[have]

````

Like most comparative methods, this analysis will require ultrametric trees
(branch lengths proportional to time, rather than to mutational steps). 
As most of these phylogenies are calibrated with branch length proportional
to mutational step, we must time-calibrate each of them first.  

``` {r timetree, cache=TRUE}
timetree <- function(tree)
    try( chronoMPL(multi2di(tree)) )
tt <- drop_nontrees(sapply(branchlengths, timetree))
````
At this point we have `r length(tt)` time-calibrated phylogenies over which
we will apply the diversification rate analysis. 
Computing the gamma test statistic to identify devations from the 
constant-rates model takes a single line,

``` {r }
gammas <- sapply(tt,  gammaStat)
````
and the resulting distribution of the statistic across available trees is shown Fig 3.

``` {r   gammadist, fig.cap="Distribution of the gamma statistic across phylogenies in TreeBASE. Strongly positive values are indicative of an increasing rate of evolution (excess of nodes near the tips), very negative values indicate an early burst of diversification (an excess of nodes near the root)."   }
qplot(gammas)
````


Because `treebase` makes it possible to perform this analysis entirely
by scripts using the latest treebase data, it is not only easier to
perform this analysis but also to update it to reflect the latest data.
Note that in this example it is not our objective to provide a thorough
analysis of diversification patterns and their possible interpretations,
as in @pybus2000, @mcpeek2007, @mcpeek2008 and @phillimore2008, but
merely to illustrate how the similar calculations to these can be easily
applied across the much larger datasets in the repository. This example
can be automatically updated to reflect the latest data in TreeBASE
simply by rerunning the code we present above.


# References

