<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title></title>
  <style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
  </style>
</head>
<body>
<p>cboettig@ucdavis.edu</p>
<p>[cor1]Corresponding author.</p>
<ol style="list-style-type: decimal">
<li><p>TreeBASE is an important and rapidly growing repository of phylogenetic data. The R statistical environment has become a primary tool for the applied phylogenetic analyses that use this kind of data for across a range of questions, from comparative evolution to community ecology to conservation planning.</p></li>
<li><p>We have developed <code>treebase</code>, an open-source package (freely available from <a href="http://cran.r-project.org/web/packages/treebase">http://cran.r-project.org/web/packages/treebase</a>) for the R environment, providing simplified, programmatic and interactive access to phylogenetic data in the TreeBASE repository.</p></li>
<li><p>We illustrate how this package creates a bridge between the repository and the rapidly growing ecosystem of R packages for phylogenetics that can reduce barriers to discovery and integration across phylogenetic research.</p></li>
<li><p>We show how the <code>treebase</code> package can be used to facilitate replication of previous studies and testing of methods and hypotheses across a large sample of phylogenies, which may help make these practices more common.</p></li>
</ol>
<h4 id="keywords">Keywords</h4>
<p>R, software, API, TreeBASE, database, e-science</p>
<h1 id="introduction">Introduction</h1>
<p>Applications that use phylogenetic information as part of their analyses are becoming increasingly central to both evolutionary and ecological research. The exponential growth in genetic sequence data available for all forms of life has driven rapid advances in the methods that can infer the phylogenetic relationships and divergence times across different taxa [@huelsenbeck2001b; @stamatakis2006; @drummond2007]. just as the availability of sequence data has led to the subsequent explosion of phylogenetic methods, and many other avenues of research, this rapid expanse of phylogenetic data now primes new innovations across ecology and evolution. Once again the product of one field has become the raw data of the next. Unfortunately, while the discipline of bioinformatics has emerged to help harness and curate the wealth of genetic data with cutting edge computer science, statistics, and internet technology, its counterpart in evolutionary informatics remains “scattered, poorly documented, and in formats that impede discovery and integration” [@parr2011a]. Our goal when developing the <code>treebase</code> package was to address this gap by providing how programmatic and interactive access between the repositories that store this data and the software tools commonly used to analyze them.</p>
<p>In this paper we focus on applications which use rather than generate phylogenetic data. Such approaches stand to benefit substantially from this programmatic and interactive access to TreeBASE. While the task of inferring phylogenies from sequence data remains dominated by dedicated compiled software such as MrBayes [@huelsenbeck2001b], BEAST [@drummond2007], RAXML [@stamatakis2006], the ever-growing suite of research methods that use these phylogenies as input data are largely based in R. The R statistical environment [@rteam2012] has become a dominant platform for researchers using phylogenetic data to address a rapidly expanding set of questions in ecological and evolutionary processes. These methods include but are not limited to ancestral state reconstruction [@paradis2004; @butler2004], diversification analysis [@paradis2004; @rabosky2006b; @harmon2008; @fitzjohn2009; @fitzjohn2010; @goldberg2011; @stadler2011], quantifying the rate and tempo of trait evolution [@butler2004; @paradis2004; @harmon2008; @hipp2010; @revell2011; @eastman2011], identifying evolutionary influences and proxies for community ecology [@webb2008; @kembel2010], performing phyloclimatic modelling [@warren2008; @evans2009b], and simulation of speciation and character evolution [@harmon2008; @stadler2011a; @boettiger2012], as well as various manipulation and visualization of phylogenetic data [@paradis2004; @schliep2010; @jombart2010; @revell2011]. A more comprehensive list of R packages by analysis type is available on the phylogenetics taskview, <a href="http://cran.r-project.org/web/views/Phylogenetics.html">http://cran.r-project.org/web/views/Phylogenetics.html</a>. Several programs exist outside the R language for applied phylogenetic methods, incuding Java [@maddison2011], MATLAB [@blomberg2003] and Python [@sukumaran2010] and online interfaces [@martins2004].</p>
<p>TreeBASE (<a href="http://treebase.org">http://treebase.org</a>) is an online repository of phylogenetic data (e.g. trees of species, populations, or genes) that have been published in a peer-reviewed academic journal, book, thesis or conference proceedings [@sanderson1994b; @morell1996]. The database can be searched through an online interface which allows users to find a phylogenetic tree from a particular publication, author or taxa of interest. TreeBASE provides an application programming interface (API) that lets computer applications make queries to the database. Our <code>treebase</code> package uses this API to create a direct link between this data and the R language. This has several immediate and important benefits:</p>
<ol style="list-style-type: decimal">
<li><p><em>Data discovery.</em> Users can leverage the rich statstical environment provided by the R language to better identify data sets appropriate for their research.</p></li>
<li><p><em>Programmatic data access.</em> Many tasks that are theoretically made possible by the creation of the TreeBASE repository are not pursued because they would be too laborious for an exploratory analysis. The ability to use loops to automatically download and perform a systematic analysis using the rich set of tools available in R opens up new avenues for research.</p></li>
<li><p><em>Automatic updating</em>. The TreeBASE repository is expanding rapidly. The scriptable nature of analyses run in with our <code>treebase</code> package means that a study can be rerun on the latest version of the repository without additional effort.</p></li>
</ol>
<h2 id="basic-queries">Basic queries</h2>
<p>The basic functions of the TreeBASE API allow search queries through two separate interfaces. The <code>OAI-PMH</code> interface provides the metadata associated with the publications from which the phylogenies have been taken, while the <code>Phylo-WS</code> interface provides information and access to the phylogenetic data itself. These interfaces are well-documented on the TreeBASE website. The <code>treebase</code> package allows these queries to be made directly from R, just as a user would make them from the browser. Because the queries can be implemented programmatically in R, a user can construct more complicated filters than permitted by the web interface, and can maintain a record of the queries they used to collect their data as an R script. The ability to script this data-gathering step of research can go a long way to reducing errors and ensuring that an analysis can be replicated later, by the author or other groups [@peng2011a].</p>
<p>Any of the queries available on the web interface can now be made directly from R, including downloading and importing the phylogeny into the R interface. For instance, one can search for phylogenies containing dolphin taxa, “Delphinus,” or all phylogenies submitted by a given author, “Huelsenbeck”,</p>
<pre class="sourceCode r"><code class="sourceCode r">    <span class="kw">search_treebase</span>(<span class="st">&quot;Delphinus&quot;</span>, <span class="dt">by=</span><span class="st">&quot;taxon&quot;</span>)
    <span class="kw">search_treebase</span>(<span class="st">&quot;Huelsenbeck&quot;</span>, <span class="dt">by=</span><span class="st">&quot;author&quot;</span>)</code></pre>
<p>This function loads the matching phylogenies into R, ready for analysis. The package documentation provides many examples of possible queries. The <code>search_treebase</code> function is the heart of the <code>treebase</code> package. Table 1 lists each of the types of queries available through the <code>search_treebase</code> function. This list can also be found in the function documentation, <code>?search_treebase</code>.</p>
<table>
<tbody>
<tr>
<th>
search “by=”
</th>
<th>
description
</th>
</tr>
<tr>
<td>
abstract
</td>
<td>
search terms in the publication abstract
</td>
</tr>
<tr>
<td>
author
</td>
<td>
match authors in the publication
</td>
</tr>
<tr>
<td>
subject
</td>
<td>
match subject
</td>
</tr>
<tr>
<td>
doi
</td>
<td>
the unique object identifier for the publication
</td>
</tr>
<tr>
<td>
ncbi
</td>
<td>
NCBI identifier number for the taxon
</td>
</tr>
<tr>
<td>
ubio
</td>
<td>
“tb.identifier.ubio”,
</td>
</tr>
<tr>
<td>
kind.tree
</td>
<td>
Kind of tree (Gene tree, species tree, barcode tree)
</td>
</tr>
<tr>
<td>
type.tree
</td>
<td>
type of tree (Consensus or Single)
</td>
</tr>
<tr>
<td>
ntax
</td>
<td>
number of taxa in the matrix
</td>
</tr>
<tr>
<td>
quality
</td>
<td>
A quality score for the tree, if it has been rated.
</td>
</tr>
<tr>
<td>
study
</td>
<td>
match words in the title of the study or publication
</td>
</tr>
<tr>
<td>
taxon
</td>
<td>
taxon scientific name
</td>
</tr>
<tr>
<td>
id.study
</td>
<td>
TreeBASE study ID
</td>
</tr>
<tr>
<td>
id.tree
</td>
<td>
TreeBASE’s unique tree identifier (Tr.id)
</td>
</tr>
<tr>
<td>
id.taxon
</td>
<td>
taxon identifier number from TreeBase
</td>
</tr>
<tr>
<td>
tree
</td>
<td>
The title for the tree
</td>
</tr>
</tbody>
</table>

<p>Table 1: Queries available in <code>search_treebase</code></p>
<h1 id="data-discovery-in-treebase">Data discovery in TreeBASE</h1>
<p>The <code>treebase</code> package provides access to the metadata of all publications containing trees deposited in TreeBASE using a separate API built on the OAI-PMH protocol, an international web standard such data. This can help the user discover phylogenies of interest and also allows the user to perform statistical analyses on the data deposition itself, which could identify trends or biases in the phylogenetics literature.</p>
<p>This publication metadata is accessed by <code>search_metadata</code> function, which can download the metadata for all publications associated with TreeBASE.</p>
<pre class="sourceCode r"><code class="sourceCode r">    oai.md &lt;- <span class="kw">search_metadata</span>() </code></pre>
<p>This returns an R list object, in which each element is an entry with bibliographic information corresponding to a published study that has deposited data in TreeBASE. From the length of this list we see that there are currently 3105 published studies in the database.</p>
<p>The <code>oai_metadata</code> function facilitates extracting the different meta-data fields. For instance, to obtain a list of all the dates of publication &amp; names of the journals (publishers) that have submitted data:</p>
<pre class="sourceCode r"><code class="sourceCode r">    dates &lt;- <span class="kw">oai_metadata</span>(<span class="st">&quot;date&quot;</span>, oai.md) 
    pub &lt;- <span class="kw">oai_metadata</span>(<span class="st">&quot;publisher&quot;</span>, oai.md)</code></pre>
<p>Many journals have only a few submissions, so we will classify any not in the top ten contributing journals as “Other”:</p>
<pre class="sourceCode r"><code class="sourceCode r">    topten &lt;- <span class="kw">sort</span>(<span class="kw">table</span>(pub), <span class="dt">decreasing=</span><span class="ot">TRUE</span>)[<span class="dv">1</span>:<span class="dv">10</span>]
    pub[!(pub %in% <span class="kw">names</span>(topten))] &lt;- <span class="st">&quot;Other&quot;</span></code></pre>
<p>We plot the distribution of publication years for phylogenies deposited in TreeBASE, color coding by publisher in Fig [fig:1].</p>
<pre class="sourceCode r"><code class="sourceCode r">    meta &lt;- <span class="kw">data.frame</span>(<span class="dt">pub =</span> pub, <span class="dt">dates =</span> dates)
    <span class="kw">ggplot</span>(meta) + <span class="kw">geom_bar</span>(<span class="kw">aes</span>(dates, <span class="dt">fill =</span> publisher))</code></pre>
<pre><code>Error: could not find function &quot;ggplot&quot;</code></pre>
<p>Typically we are more interested in the metadata describing the phylogenies themselves rather than the publications in which they appeared, such as the number of taxa in the tree, a quality score (if available), kind of tree (gene tree, species tree, or barcode tree) or whether the phylogeny represents a consensus tree from a distribution or just a single estimate. The <code>cache_treebase</code> function is used to download all available phylogenies from TreeBASE. Here, we call the function with an optional argument that will return only the metadata just listed for all available phylogenies, which runs much more quickly.</p>
<pre class="sourceCode r"><code class="sourceCode r">    phylo.md &lt;- <span class="kw">cache_treebase</span>(<span class="dt">only_metadata=</span><span class="ot">TRUE</span>)</code></pre>
<p>We can summarize how these 10,555 trees break out by kind or type (The <code>xtable</code> command formats this as a table) using the <code>phylo_metadata</code> function to extract the kind of tree (gene/species/barcode) and type (single or consensus):</p>
<pre class="sourceCode r"><code class="sourceCode r"> output &lt;- <span class="kw">table</span>(<span class="kw">phylo_metadata</span>(<span class="st">&quot;kind&quot;</span>, phylo.md), <span class="kw">phylo_metadata</span>(<span class="st">&quot;type&quot;</span>, phylo.md))
 xtable::<span class="kw">xtable</span>(output)</code></pre>
<!-- html table generated in R 2.15.0 by xtable 1.7-0 package -->
<!-- Sat May 12 14:33:29 2012 -->
<TABLE border=1>
<TR> <TH>  </TH> <TH> 
Consensus
</TH> <TH> 
Single
</TH>  </TR>
  <TR> <TD align="right"> 
Barcode Tree
</TD> <TD align="right">   
1
</TD> <TD align="right">  
11
</TD> </TR>
  <TR> <TD align="right"> 
Gene Tree
</TD> <TD align="right"> 
123
</TD> <TD align="right"> 
214
</TD> </TR>
  <TR> <TD align="right"> 
Species Tree
</TD> <TD align="right"> 
2952
</TD> <TD align="right"> 
7239
</TD> </TR>
   </TABLE>




<p>For certain applications a user may wish to download all the available phylogenies from TreeBASE. Using the <code>cache_treebase</code> function allows a user to download a local copy of all trees. Because direct database dumps are not available, this function has intentional delays to avoid overtaxing the TreeBASE servers, and should be allowed a full day to run.</p>
<pre class="sourceCode r"><code class="sourceCode r">    treebase &lt;- <span class="kw">cache_treebase</span>()</code></pre>
<p>Once run, the cache is saved compactly in memory where it can be easily and quickly restored. For convenience, the <code>treebase</code> package comes with a copy already cached, which can be loaded into memory.</p>
<pre class="sourceCode r"><code class="sourceCode r">    <span class="kw">data</span>(treebase)</code></pre>
<p>Having access to both the metadata from the studies and from the phylogenies in R lets us quickly combine these data sources in interesting ways. For instance, with a few commands we can visualize how the number of taxa on submitted phylogenies has increasing over time, Figure [fig:2].</p>
<pre class="sourceCode r"><code class="sourceCode r">    phylo.id &lt;- <span class="kw">phylo_metadata</span>(<span class="st">&quot;Study.id&quot;</span>, phylo.md)
    oai.id &lt;- <span class="kw">oai_metadata</span>(<span class="st">&quot;Study.id&quot;</span>, oai.md)
    matches &lt;- <span class="kw">sapply</span>(oai.id, match, phylo.id)
    Ntaxa &lt;- <span class="kw">phylo_metadata</span>(<span class="st">&quot;ntaxa&quot;</span>,  phylo.md[matches])
    Ntaxa[<span class="kw">sapply</span>(Ntaxa, is.null)] &lt;- <span class="ot">NA</span>
    taxa &lt;- <span class="kw">data.frame</span>(<span class="dt">Ntaxa=</span><span class="kw">as.numeric</span>(<span class="kw">unlist</span>(Ntaxa)), meta)
    <span class="kw">ggplot</span>(taxa, <span class="kw">aes</span>(dates, Ntaxa)) + 
      <span class="kw">geom_point</span>(<span class="dt">position =</span> <span class="st">&#39;jitter&#39;</span>, <span class="dt">alpha =</span> .<span class="dv">8</span>) + 
      <span class="kw">scale_y_log10</span>() + <span class="kw">stat_smooth</span>(<span class="kw">aes</span>(<span class="dt">group =</span> <span class="dv">1</span>))</code></pre>
<div class="figure">
<img src="http://farm8.staticflickr.com/7242/7184238818_6564ca0e1c_o.png" alt="Combining the metadata available from publications and from phylogenies themselves, we can visualize the growth in taxa on published phylogenies. Note that the maximum size tree deposited each year is growing far faster than the average number." /><p class="caption">Combining the metadata available from publications and from phylogenies themselves, we can visualize the growth in taxa on published phylogenies. Note that the maximum size tree deposited each year is growing far faster than the average number.</p>
</div>
<p>The promise of this exponential growth in the sizes of available phylogenies, with some trees representing 2,957 taxa motivates the more and more ambitious inference methods being developed which require large trees to have adequate signal [@boettiger2012; @fitzjohn2009; @beaulieu2012].</p>
<h1 id="reproducible-research">Reproducible research</h1>
<p>Reproducible research has become a topic of increasing concern in recent years, and facilitating access to data and using scripts that can replicate analyses can help lower barriers to the replication of statistical and computational results [@schwab2000; @gentleman2004; @peng2011b]. The <code>treebase</code> package facilitates this process, as we illustrate in a simple example.</p>
<p>Consider the shifts in speciation rate identified by @derryberry2011 on a phylogeny of ovenbirds and treecreapers, which uses the methods provided in the R package <code>laser</code> [@rabosky2006b]. We will seek to not only replicate the results, but also compare them against methods presented in @stadler2011 in the package <code>TreePar</code>, which permits speciation models that were not available to @derryberry2011 at the time of their study.</p>
<h2 id="obtaining-the-tree">Obtaining the tree</h2>
<p>By drawing on the rich data manipulation tools available in R which should be familiar to the large R phylogenetics community, the <code>treebase</code> package allows us to construct richer queries than are possible through TreeBASE alone.</p>
<p>The fastest way to identify the data uses the digital object identifer (doi) printed on the top of the article,</p>
<pre class="sourceCode r"><code class="sourceCode r">results &lt;- <span class="kw">search_treebase</span>(<span class="st">&quot;10.1111/j.1558-5646.2011.01374.x&quot;</span>, <span class="st">&quot;doi&quot;</span>)</code></pre>
<p>The search returns a list, since some studies contain many trees</p>
<pre class="sourceCode r"><code class="sourceCode r">results </code></pre>
<pre><code>[[1]]

Phylogenetic tree with 297 tips and 296 internal nodes.

Tip labels:
    Formicarius_colma, Glyphorynchus_spirurus, Dendrocolaptes_platyrostris, Dendrocolaptes_picumnus, Dendrocolaptes_hoffmannsi, Dendrocolaptes_certhia_concolor, ...

Rooted; includes branch lengths.
</code></pre>
<p>In this case the single item in the list is the one we want:</p>
<pre class="sourceCode r"><code class="sourceCode r">derryberry &lt;- results[[<span class="dv">1</span>]]</code></pre>
<p>Having successfully imported the phylogenetic tree corresponding to this study, we can quickly replicate their analysis of which diversification process best fits the data. These steps can be easily implemented using the phylogenetics packages we have just mentioned. As a comparison of speciation models is not the focus of this paper, the complete code and explanation for these steps is provided as an appendix. Happily, this analysis confirms the author’s original conclusions, even when the more general model of @stadler2011 is considered.</p>
<h1 id="analyses-across-many-phylogenies">Analyses across many phylogenies</h1>
<p>Large scale comparative analyses that seek to characterize evolutionary patterns across many phylogenies increasingly common in phylogenetic methods [<em>e.g.</em> @mcpeek2007; @phillimore2008; @mcpeek2008; @quental2010; @davies2011a].<br />Sometimes referred to by their authors as meta-analyses, these approaches have focused on re-analyzing phylogenetic trees collected from many different earlier publications.<br />This is a more direct approach than the traditional concept of meta-analysis where statistical results from earlier studies are weighted by their sample size without actually repeating the statistical analyses of those papers. Because the identical analysis can be repeated on the original data from each study, this approach avoids some of the statistical challenges inherent in traditional meta-analyses summarizing results across heterogeneous approaches.</p>
<p>To date, researchers have gone through heroic efforts simply to assemble these data sets from the literature.<br />As described in @mcpeek2007; (emphasis added) &gt; One data set was based on 163 published species-level molecular phylogenies &gt; of arthropods, chordates, and mollusks. A PDF format file of each article &gt; was obtained, and a digital snapshot of the figure was taken in Adobe Acrobat &gt; 7.0. This image was transferred to a PowerPoint (Microsoft) file and printed &gt; on a laser printer. The phylogenies included in this study are listed in the &gt; appendix. <em>All branch lengths were measured by hand from these printed sheets &gt; using dial calipers.</em></p>
<p>Despite the recent appearance of digital tools that could now facilitate this analysis without mechanical calipers, [<em>e.g.</em> treesnatcher, @laubach2007], it is easier and less error-prone to pull properly formatted phylogenies from the database for this purpose. Moreover, as the available data increases with subsequent publications, updating earlier meta-analyses can become increasingly tedious. Using our package, a user can apply any analysis they have written for a single phylogeny across the entire collection of suitable phylogenies in TreeBASE, which can help overcome such barriers to discovery and integration at this large scale. Using the functions we introduce aboved, we provide in the Appendix a simple example that computes the gamma statistic of @pybus2000, which provides an measure of when speciation patterns differ from the popular birth-death model. We see that many phylogenies fall outside the distribution of the statistic expected under that model.</p>
<h1 id="conclusion">Conclusion</h1>
<p>While we have focused on examples that require no additional data beyond the phylogeny, a wide array of methods combine this data with information about the traits, geography, or ecological community of the taxa represented. In such cases we would need programmatic access to the trait data as well as the phylogeny. The Dryad digital repository (<a href="http://datadryad.org">http://datadryad.org</a>) is a popular host for such data to support the data archiving requirements mentioned above. While programmatic access to the repository is possible through the <code>rdryad</code> package [@chamberlain2012], variation in data formatting must first be overcome. Dedicated databases such as FishBASE (<a href="http://fishbase.org">http://fishbase.org</a>) may be another alternative, where morphological data can be queried for a list of species using the <code>rfishbase</code> package <span class="citation">(Boettiger)</span>. The development of similar software for programmatic data access will rapidly extend the space and scale of possible analyses.</p>
<p>The recent advent of mandatory data archiving in many of the major journals publishing phylognetics-based research [<em>e.g.</em> @fairbairn2010; @piwowar2011; @whitlock2010], is a particularly promising development that should continue to fuel trend of submissions seen in Fig. 1. Accompanied by faster and more inexpensive techniques of NextGen sequencing, and the rapid expansion in phylogenetic applications, we anticipate this rapid growth in available phylogenies will continue. Faced with this flood of data, programmatic access becomes not only increasingly powerful but an increasingly necessary way to ensure we can still see the forest for all the trees.</p>
<h1 id="acknowledgements">Acknowledgements</h1>
<p>CB wishes to thank S. Price for feedback on the manuscript, the TreeBASE developer team for building and supporting the repository, and all contributers to TreeBASE. CB is supported by a Computational Sciences Graduate Fellowship from the Department of Energy under grant number DE-FG02-97ER25308.</p>
<h1 id="references">References</h1>
<h1 id="appendix">Appendix</h1>
<h2 id="reproducible-research-a-diversification-rate-analysis">Reproducible Research: A diversification rate analysis</h2>
<p>Different diversification models make different assumptions about the rate of speciation, extinction, and how these rates may be changing over time. The authors consider eight different models, implemented in the laser package [@rabosky2006b]. This code fits each of the eight models to that data:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ape)
bt &lt;- <span class="kw">branching.times</span>(derryberry)
<span class="kw">library</span>(laser)
models &lt;- <span class="kw">list</span>(             <span class="dt">yule =</span> <span class="kw">pureBirth</span>(bt),  
                     <span class="dt">birth_death =</span> <span class="kw">bd</span>(bt),     
                     <span class="dt">yule.2.rate =</span> <span class="kw">yule2rate</span>(bt),
      <span class="dt">linear.diversity.dependent =</span> <span class="kw">DDL</span>(bt),    
 <span class="dt">exponential.diversity.dependent =</span> <span class="kw">DDX</span>(bt),
         <span class="dt">varying.speciation_rate =</span> <span class="kw">fitSPVAR</span>(bt),  
         <span class="dt">varying.extinction_rate =</span> <span class="kw">fitEXVAR</span>(bt),  
                    <span class="dt">varying_both =</span> <span class="kw">fitBOTHVAR</span>(bt)  
              )</code></pre>
<p>Each of the model estimate includes an AIC score indicating the goodness of fit, penalized by model complexity (lower scores indicate better fits) We ask R to tell us which model has the lowest AIC score,</p>
<pre class="sourceCode r"><code class="sourceCode r">aics &lt;- <span class="kw">sapply</span>(models, <span class="st">`</span><span class="dt">[[</span><span class="st">`</span>, <span class="st">&#39;aic&#39;</span>)
best_fit &lt;- <span class="kw">names</span>(models[<span class="kw">which.min</span>(aics)])</code></pre>
<p>and confirm the result presented in @derryberry2011; that the yule.2.rate model is the best fit to the data.</p>
<p>In this fast-moving field, new methods often become available within the time-frame that another manuscript is submitted by its authors and the time at which if first appears in print. For instance, the more sophisticated methods available in the more recent package, <code>TreePar</code>, introduced in @stadler2011 were not used in this study.</p>
<p>We load the new method and format the data as its manual instructs us</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(TreePar)
x&lt;-<span class="kw">sort</span>(<span class="kw">getx</span>(derryberry), <span class="dt">decreasing=</span><span class="ot">TRUE</span>)</code></pre>
<p>The best-fit model in the laser analysis was a yule (net diversification rate) models with two separate rates. We can ask <code>TreePar</code> to see if a model with more rate shifts is favored over this single shift, a question that was not possible to address using the tools provided in <code>laser</code>. The previous analysis also considers a birth-death model that allowed speciation and extinction rates to be estimated separately, but did not allow for a shift in the rate of such a model. Here we consider models that have up to 4 different rates in Yule models, (The syntax in <code>TreeParr</code> is slightly cumbersome, the [[2]] indicates where this command happens to store the output models.)</p>
<pre class="sourceCode r"><code class="sourceCode r">yule_models &lt;- <span class="kw">bd.shifts.optim</span>(x, <span class="dt">sampling =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>), 
  <span class="dt">grid =</span> <span class="dv">5</span>, <span class="dt">start =</span> <span class="dv">0</span>, <span class="dt">end =</span> <span class="dv">60</span>, <span class="dt">yule =</span> <span class="ot">TRUE</span>)[[<span class="dv">2</span>]]</code></pre>
<p>We also want to compare the performance of models which allow up to four shifts and also estimate extinction and speciation separately:</p>
<pre class="sourceCode r"><code class="sourceCode r">birth_death_models &lt;- <span class="kw">bd.shifts.optim</span>(x, <span class="dt">sampling =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>), 
  <span class="dt">grid =</span> <span class="dv">5</span>, <span class="dt">start =</span> <span class="dv">0</span>, <span class="dt">end =</span> <span class="dv">60</span>, <span class="dt">yule =</span> <span class="ot">FALSE</span>)[[<span class="dv">2</span>]]</code></pre>
<p>The models output by these functions are ordered by increasing number of shifts.<br />We can select the best-fitting model by AIC score,</p>
<pre class="sourceCode r"><code class="sourceCode r">yule_aic &lt;- <span class="kw">sapply</span>(yule_models, function(pars) <span class="dv">2</span> * (<span class="kw">length</span>(pars) - <span class="dv">1</span>) + <span class="dv">2</span> * pars[<span class="dv">1</span>] )
birth_death_aic &lt;- <span class="kw">sapply</span>(birth_death_models, function(pars) <span class="dv">2</span> * (<span class="kw">length</span>(pars) - <span class="dv">1</span>) + <span class="dv">2</span> * pars[<span class="dv">1</span>] )
best_no_of_rates &lt;- <span class="kw">list</span>(<span class="dt">Yule =</span> <span class="kw">which.min</span>(yule_aic), <span class="dt">birth.death =</span> <span class="kw">which.min</span>(birth_death_aic))
best_model &lt;- <span class="kw">which.min</span>(<span class="kw">c</span>(<span class="kw">min</span>(yule_aic), <span class="kw">min</span>(birth_death_aic)))</code></pre>
<p>which confirms that the Yule 2-rate<br />model is still the best choice based on AIC score.<br />Of the eight models in this second analysis, only three were in the original set considered (Yule 1-rate and 2-rate, and birth-death without a shift), so we could by no means have been sure ahead of time that a birth death with a shift, or a Yule model with a greater number of shifts, would not have fitted better.</p>
<table>
<thead>
<tr class="header">
<th align="left">Testing for constant speciation and extinction rates across many phylogenies</th>
</tr>
</thead>
<tbody>
</tbody>
</table>
<p>A standard test of this is the gamma statistic of @pybus2000 which tests the null hypothesis that the rates of speciation and extinction are constant. The gamma statistic is normally distributed about 0 for a pure birth or birth-death process, values larger than 0 indicate that internal nodes are closer to the tip then expected, while values smaller than 0 indicate nodes farther from the tip then expected. In this section, we collect all phylogenetic trees from TreeBASE and select those with branch length data that we can time-calibrate using tools available in R. We can then calculate the distribution of this statistic for all available trees, and compare these results with those from the analyses mentioned above.</p>
<p>The <code>treebase</code> package provides a compressed cache of the phylogenies available in treebase. This cache can be automatically updated with the <code>cache_treebase</code> function,</p>
<pre class="sourceCode r"><code class="sourceCode r">treebase &lt;- <span class="kw">cache_treebase</span>()</code></pre>
<p>which may require a day or so to complete, and will save a file in the working directory named with treebase and the date obtained. For convenience, we can load the cached copy distributed with the <code>treebase</code> package:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(treebase)</code></pre>
<p>We will only be able to use those phylogenies that include branch length data. We drop those that do not from the data set,</p>
<pre class="sourceCode r"><code class="sourceCode r">      have &lt;- <span class="kw">have_branchlength</span>(treebase)
      branchlengths &lt;- treebase[have]</code></pre>
<p>Like most comparative methods, this analysis will require ultrametric trees (branch lengths proportional to time, rather than to mutational steps). As most of these phylogenies are calibrated with branch length proportional to mutational step, we must time-calibrate each of them first.</p>
<pre class="sourceCode r"><code class="sourceCode r">timetree &lt;- function(tree){
    <span class="kw">try</span>( <span class="kw">chronoMPL</span>(<span class="kw">multi2di</span>(tree)) )
}
tt &lt;- <span class="kw">drop_nontrees</span>(<span class="kw">sapply</span>(branchlengths, timetree))</code></pre>
<p>At this point we have 1,217 time-calibrated phylogenies over which we will apply the diversification rate analysis. Computing the gamma test statistic to identify devations from the constant-rates model takes a single line,</p>
<pre class="sourceCode r"><code class="sourceCode r">gammas &lt;- <span class="kw">sapply</span>(tt,  gammaStat)</code></pre>
<p>and the resulting distribution of the statistic across available trees is shown Fig 3.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">qplot</span>(gammas)</code></pre>
<div class="figure">
<img src="http://farm9.staticflickr.com/8025/7184300690_57eb8fff25_o.png" alt="plot of chunk gammadist" /><p class="caption">plot of chunk gammadist</p>
</div>
<p>Because <code>treebase</code> makes it possible to perform this analysis entirely by scripts using the latest treebase data, it is not only easier to perform this analysis but also to update it to reflect the latest data. Note that in this example it is not our objective to provide a thorough analysis of diversification patterns and their possible interpretations, as in @pybus2000; @mcpeek2007; @mcpeek2008; and @phillimore2008; but merely to illustrate how the similar calculations to these can be easily applied across the much larger datasets in the repository. This example can be automatically updated to reflect the latest data in TreeBASE simply by rerunning the code we present above.</p>
<p>Boettiger, Carl. “rfishbase: R Interface to FishBASE.”</p>
</body>
</html>
